pandas/core/groupby/generic.py

```diff
@@ -1664,3 +1655,36 @@ def _normalize_keyword_aggregation(kwargs):
         order.append((column,
                       com.get_callable_name(aggfunc) or aggfunc))
     return aggspec, columns, order
+
+
+def _recast_datetimelike_result(result):
+    """
+    If we have date/time like in the original, then coerce dates
+    as we are stacking can easily have object dtypes here.
+
+    Parameters
+    ----------
+    result : DataFrame
+
+    Notes
+    -----
+    - Assumes Groupby._selected_obj has ndim==2 and at least one
+    datetimelike column
+    - Modifies `result` inplace
+    """
+    ocols = [idx for idx in range(len(result.columns))
+             if result.dtypes[idx] == object]
+
+    for cidx in ocols:
```

Comment:
this routines feels almost like maybe_convert_objects

Reply:
Ha yep, in the next line there's a comment "# TODO: get maybe_convert_objects working here".  I tried this and got shiny new failures bc maybe_convert_objects expects a 1D ndarray.

Reply:
maybe_convert_objects is only 1d, my point is to *use* it here

Reply:
When I do this, I get one test failure, but it looks like it may be another test-is-wrong thing, need confirmation.

tests.groupby.test_apply.test_apply_numeric_coercion_when_datetime

```
df = pd.DataFrame({'Number': [1, 2],
                           'Date': ["2017-03-02"] * 2,
                           'Str': ["foo", "inf"]})

        expected = df.groupby(['Number']).apply(lambda x: x.iloc[0])
        df.Date = pd.to_datetime(df.Date)
        result = df.groupby(['Number']).apply(lambda x: x.iloc[0])
        tm.assert_series_equal(result['Str'], expected['Str'])
```

In master, `result['Date']` has object-dtype.  If I use maybe_convert_objects here, it has datetime64[ns] dtype.  Should it be object dtype?

Reply:
Can you double-check? On master I have `result.dtypes`


```python
In [10]: result.dtypes
Out[10]:
Number             int64
Date      datetime64[ns]
Str               object
dtype: object
```

Reply:
Good catch, I had this one wrong in a couple of different directions.